src/jit/gentree.h

```diff
@@ -4626,24 +4636,58 @@ struct GenTreeAllocObj final : public GenTreeUnOp
 // be defined already.
 //------------------------------------------------------------------------
 
-inline bool GenTree::OperIsBlkOp() const
+inline bool GenTree::OperIsBlkOp()
 {
-    return (gtOper == GT_INITBLK || gtOper == GT_COPYBLK || gtOper == GT_COPYOBJ);
+    return (((gtOper == GT_ASG) && varTypeIsStruct(gtOp.gtOp1))
+#ifndef LEGACY_BACKEND
+            || (OperIsBlk() && (AsBlk()->Data() != nullptr))
+#endif
+                );
 }
 
 inline bool GenTree::OperIsDynBlkOp()
 {
-    return (OperIsBlkOp() && !gtGetOp2()->IsCnsIntOrI());
+    if (gtOper == GT_ASG)
+    {
+        return gtGetOp1()->OperGet() == GT_DYN_BLK;
+    }
+#ifndef LEGACY_BACKEND
+    else if (gtOper == GT_STORE_DYN_BLK)
+    {
+        return true;
+    }
+#endif
+    return false;
 }
 
-inline bool GenTree::OperIsCopyBlkOp() const
+inline bool GenTree::OperIsCopyBlkOp()
 {
-    return (gtOper == GT_COPYOBJ || gtOper == GT_COPYBLK);
+    if (gtOper == GT_ASG)
+    {
+        return (varTypeIsStruct(gtGetOp1()) && ((gtFlags & GTF_BLK_INIT) == 0));
+    }
+#ifndef LEGACY_BACKEND
+    else if (OperIsBlk())
+    {
+        return ((AsBlk()->Data() != nullptr) && ((gtFlags & GTF_BLK_INIT) == 0));
+    }
+#endif
+    return false;
 }
 
-inline bool GenTree::OperIsInitBlkOp() const
+inline bool GenTree::OperIsInitBlkOp()
 {
-    return (gtOper == GT_INITBLK);
+    if (gtOper == GT_ASG)
+    {
+        return (varTypeIsStruct(gtGetOp1()) && ((gtFlags & GTF_BLK_INIT) != 0));
+    }
+#ifndef LEGACY_BACKEND
+    else if (OperIsBlk())
+    {
+        return ((AsBlk()->Data() != nullptr) && ((gtFlags & GTF_BLK_INIT) != 0));
```

Comment:
Ah, I see. I missed that you were suggesting to eliminate the `OperIsBlk()` check (rather than just replace the check on Data. I'll make that change.

---
In reply to: [77390490](https://github.com/dotnet/coreclr/pull/7027#discussion_r77390490) [](ancestors = 77390490,77387651,77381269)